_posts/2023-07-08-our-infrastructure.md

```diff
@@ -0,0 +1,261 @@
+Either way, slapping a simple Cloudflare Worker between it really isn't the biggest problem.
+
+### Cloudflare Workers
+
+[Cloudflare Workers](https://workers.cloudflare.com/) allows running small pieces of Javascript or WASM in the Cloudflare network, to handle HTTP(S) traffic.
+It can do all kinds of things, like serving files from Cloudflare R2, redirect traffic, cache pages, and much more.
+
+In that sense they are very similar to AWS Lambdas, just way more powerful.
+
+Almost all our websites have a Cloudflare Worker in front of them to handle one thing or the other.
+For example, when you visit [https://forum.openttd.org](https://forum.openttd.org) you end up on a Cloudflare Worker, which returns a redirect message.
+But also our [CDN](https://cdn.openttd.org) makes use of a Cloudflare Worker: R2 (and any bucket service for that matter) normally doesn't serve `index.html` when you visit a folder.
+But to give the CDN the look and feel of a directory listing, it is important that it does.
+This is what the Cloudflare Worker is taking care of.
+If you visit a folder, it actually requests the `index.html` and serves that; it looks to you like it is listing a directory, but in fact it is just serving a HTML file.
+
+### Nomad
+
+[Nomad](https://www.nomadproject.io/) is similar to Kubernetes, AWS ECS, AWS EKS, Azure AKS, etc, but a bit simpler to work with in a day-to-day.
+Where in Kubernetes you can do everything, Nomad is more opinionated, and because of that does a lot more things for you.
+
+A nomad cluster consists of different nodes, or Nomad clients.
+A client can run jobs, which either can be a script, or a container, or really anything.
+
+In total we run five Nomad clients for OpenTTD:
+- 2 public facing clients.
+- 3 internal clients.
+
+All clients are hosted on [AWS EC2 T4g](https://aws.amazon.com/ec2/instance-types/t4/) instance, the cheapest EC2 instances AWS has.
+This type is an ARM64.
+
+To keep the AWS infrastructure as cheap as possible, we wanted to avoid needing a NAT gateway: if you use IPv4, you need something that allows you to talk with the outside world.
+On AWS you do this by installing NAT gateways.
+Sadly, those are (relatively speaking) rather expensive.
+So instead, we run as much as we can IPv6-only.
+This allows us to use an egress-only gateway, which avoids the need for a NAT gateway.
+There are a few issues with using IPv6-only, but we will touch on that in a second.
+
+Ideally we would do without the two public facing clients, but this is not possible (yet).
+Although we can route all our HTTP(S) traffic via Cloudflare (see below), our custom TCP protocols can not.
+
+To solve this, there are two solutions:
+- Either use an AWS service (read: AWS's NLB), which costs 30 USD a month.
+- Or build a Network Load Balancer yourself.
+
+We picked the second option here, and run our own NLB on the two public facing clients.
+Nomad made this painfully simple, and took only a few to set up properly.
```

Comment:
"a few" of what? Minutes, hours, seconds, days?

Reply:
Deliberately left undefined, as it isn't important to the story :)